readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/offset/HdfsOffsetComputer.java

```diff
@@ -45,51 +48,87 @@ public HdfsOffsetComputer(FileSystem fs, Path basePath, @Nullable String kafkaCl
-        String dirsPattern = String.format("%s/**", computeDirNamesPattern(LocalDateTime.now(), backlogDays));
-        FileStatus[] fileStatuses = fs.globStatus(new Path(basePath, dirsPattern));
-
-        for (FileStatus status : fileStatuses) {
-            String fileName = status.getPath().getName();
-            Matcher matcher = offsetFilePatternGenerator.matcher(fileName);
-            if (matcher.matches()) {
-                try {
-                    int partitionId = Integer.parseInt(matcher.group("partitionId"));
-                    Long offset = Long.parseLong(matcher.group("offset"));
-                    if (partitionIdsSet.contains(partitionId)) {
-                        result.merge(partitionId, offset, Long::max);
+        final Map<Integer, Map<String, FinalEventPartitionFile>> resultFile = new HashMap<>();
+
+        LocalDateTime today = LocalDateTime.now();
+        for (int i = 0; i < backlogDays; ++i) {
+            LocalDateTime day = today.minusDays(i);
+            String listedDay = day.format(DateTimeFormatter.ISO_DATE);
+
+            String dirsPattern = String.format("%s/*", computeDirName(day));
+            FileStatus[] fileStatuses = fs.globStatus(new Path(basePath, dirsPattern));
+
+            for (FileStatus status : fileStatuses) {
+                String fileName = status.getPath().getName();
+                Matcher matcher = offsetFilePatternGenerator.matcher(fileName);
+                if (matcher.matches()) {
+                    try {
+                        int partitionId = Integer.parseInt(matcher.group("partitionId"));
+                        Long index = Long.parseLong(matcher.group("index"));
+                        if (partitionIdsSet.contains(partitionId) &&
+                                (!resultFile.containsKey(partitionId) || !resultFile.get(partitionId).containsKey(listedDay) ||
+                                        index > resultFile.get(partitionId).get(listedDay).getIndex())) {
+                            HashMap<String, FinalEventPartitionFile> dateFinalEventPartitionFile =
+                                    (HashMap<String, FinalEventPartitionFile>) resultFile.computeIfAbsent(partitionId, s -> new HashMap<>());
+                            dateFinalEventPartitionFile.put(listedDay,
+                                    new FinalEventPartitionFile(index, status.getPath()));
+                            resultFile.put(partitionId, dateFinalEventPartitionFile);
+                        }
+                    } catch (NumberFormatException e) {
+                        LOGGER.info("Couldn't deviate a valid offset from '{}'", fileName);
                     }
-                } catch (NumberFormatException e) {
-                    LOGGER.info("Couldn't deviate a valid offset from '{}'", fileName);
                 }
             }
         }
 
+        // Get last offset
+        for (int partitionId : resultFile.keySet()) {
```

Comment:
could we process offsets right away and avoid the use of the intermediate hashmap ?

Otherwise, if you want to keep the hashmap, you can iterate on hashmap entrySet instead of using keySet + then getting again from the map with the key.

Reply:
I use hashmap to only read last index file per day and not to read offset on each files
So seems for me necessary to use it. Let me know if you think about something else.